builder/beacon_client.go

```diff
@@ -51,14 +51,36 @@ func (b *BeaconClient) isValidator(pubkey PubkeyHex) bool {
 }
 
 func (b *BeaconClient) getProposerForNextSlot(requestedSlot uint64) (PubkeyHex, error) {
-	/* Only returns proposer if requestedSlot is currentSlot + 1, would be a race otherwise */
+	b.updateValidatorsMap(requestedSlot)
```

Comment:
Calling `fetchEpochProposersMap` as blocking could result in the lock and the other functions taking too long.
The proposers' map is known an epoch in advance, so you could try fetching it every half epoch to avoid corner cases around epoch change.